### PR TITLE
listener: ListenConfig::listenSocketFactory

### PR DESCRIPTION
**Pull request by danzh2010**
Add ListenSocketFactory interface. Replace ListenConfig::socket() interface with listenSocketFactory() to prepare for SO_REUSEPORT.

The old TcpListenSocket creation is changed to creating a TcpListenSocketFactory which creates a TCP socket during construction in addOrUpdateListener(). This socket factory always provides the same socket when ListenSocketFactory::createListenSocket() is called.

The old UdpListenSocket create is changed to creating a UdpListenSocketFactory in addOrUpdateListener() without creating a UDP socket. The socket will be created when actual UdpListener is created in each worker thread. As currently, we don't support SO_REUSEPORT on UDP socket, createListenSocket() can only be called once.

Risk Level: low, mainly code refactoring, no change to TCP socket creation time.
Testing: existing tests

Part of #8794,  #2557

### CODE REVIEW

**Comment by mattklein123** [comment]
@danzh2010 needs a master merge after my UDP proxy PR. Can you merge master and then I will take a look?

/wait

**Comment by danzh2010** [comment]
> @danzh2010 needs a master merge after my UDP proxy PR. Can you merge master and then I will take a look?
> 
> /wait

done. PTAL!

File: source/server/hot_restarting_parent.cc [hunk 1]
```diff
@@ -95,9 +97,12 @@ HotRestartingParent::Internal::getListenSocketsForChild(const HotRestartMessage:
   Network::Address::InstanceConstSharedPtr addr =
       Network::Utility::resolveUrl(request.pass_listen_socket().address());
   for (const auto& listener : server_->listenerManager().listeners()) {
-    if (*listener.get().socket().localAddress() == *addr && listener.get().bindToPort()) {
-      wrapped_reply.mutable_reply()->mutable_pass_listen_socket()->set_fd(
-          listener.get().socket().ioHandle().fd());
+    Network::ListenSocketFactory& socket_factory = listener.get().listenSocketFactory();
+    if (*socket_factory.localAddress() == *addr && listener.get().bindToPort()) {
+      if (socket_factory.sharedSocket().has_value()) {
```
Comment (mattklein123) [inline source/server/hot_restarting_parent.cc]: Can you add a small comment that we only do socket passing if the listener socket is shared between workers and processes?

File: source/server/listener_impl.cc [hunk 2]
```diff
@@ -26,14 +26,81 @@
...
+
+Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApplyOptions() {
+  // socket might be nullptr depends on factory_ implementation.
+  Network::SocketSharedPtr socket =
+      factory_.createListenSocket(local_address_, socket_type_, options_, bind_to_port_);
+  // Binding is done by now.
+  ENVOY_LOG(info, "Create listen socket for listener {} on address {}", listener_name_,
+            local_address_->asString());
+  if (socket != nullptr && options_ != nullptr) {
+    bool ok = Network::Socket::applyOptions(options_, *socket,
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: nit: const

File: source/server/listener_impl.cc [hunk 3]
```diff
@@ -26,14 +26,81 @@
...
+
+UdpListenSocketFactory::UdpListenSocketFactory(
+    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
+    const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
+    const std::string& listener_name)
+    : ListenSocketFactoryImplBase(factory, local_address, Network::Address::SocketType::Datagram,
+                                  options, bind_to_port, listener_name) {}
+
+Network::SocketSharedPtr UdpListenSocketFactory::createListenSocket() {
+  Network::SocketSharedPtr socket = createListenSocketAndApplyOptions();
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: Can add a quick TODO here that this will fail now with more than 1 worker and there will be a follow up to use SO_REUSEPORT?

File: source/server/listener_impl.h [hunk 4]
```diff
@@ -18,6 +19,69 @@ namespace Server {
...
+  TcpListenSocketFactory(ListenerComponentFactory& factory,
+                         Network::Address::InstanceConstSharedPtr local_address,
+                         const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
+                         const std::string& listener_name);
+
+  // Network::ListenSocketFactory
+  // If |socket_| is nullptr, create a new socket for it. Otherwise, always return |socket_|.
+  Network::SocketSharedPtr createListenSocket() override;
+  absl::optional<std::reference_wrapper<Network::Socket>> sharedSocket() const override {
+    return *socket_;
```
Comment (mattklein123) [inline source/server/listener_impl.h]: ASSERT(socket != nullptr) with comment that right now all TCP sockets are shared?

File: source/server/listener_impl.h [hunk 5]
```diff
@@ -18,6 +19,69 @@ namespace Server {
...
+  }
+
+private:
+  // This will be shared across all worker threads if not setting SO_REUSEPORT.
```
Comment (mattklein123) [inline source/server/listener_impl.h]: Can you rephrase a bit just to make it clear this is not supported today? (reformat to TODO, etc.)

File: source/server/listener_manager_impl.cc [hunk 6]
```diff
@@ -473,21 +476,29 @@ void ListenerManagerImpl::drainListener(ListenerImplPtr&& listener) {
...
-        // closing it.
-        const auto& socket = listener.listener_->getSocket();
-        if (!hasListenerWithSocket(active_listeners_, socket) &&
-            !hasListenerWithSocket(warming_listeners_, socket)) {
-          // Close the socket iff it is not used anymore.
-          listener.listener_->socket().close();
+  stopListener(
+      *draining_it->listener_,
+      [this,
+       share_socket = draining_it->listener_->listenSocketFactory().sharedSocket().has_value(),
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: It's a little odd to me that are are capturing share_socket (nit should be shared_socket) here, only to return on it in the below lambda. Can you just determine this from within the lambda since you look up the listener anyway? More comments would be better also. Same below.

Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: The lambda is capturing a bool rather than the socket. The early return can avoid matching listener_tag if the socket is not shared. 

Added more comment about the share_socket.

File: source/server/connection_handler_impl.cc [hunk 7]
```diff
@@ -29,7 +29,7 @@ void ConnectionHandlerImpl::decNumConnections() {
 
 void ConnectionHandlerImpl::addListener(Network::ListenerConfig& config) {
   ActiveListenerDetails details;
-  if (config.socket().socketType() == Network::Address::SocketType::Stream) {
+  if (config.listenSocketFactory().socketType() == Network::Address::SocketType::Stream) {
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/connection_handler_impl.cc]: OK sounds good.

File: source/server/http/admin.h [hunk 8]
```diff
@@ -322,6 +322,27 @@ class AdminImpl : public Admin,
...
+    AdminListenSocketFactory(Network::SocketSharedPtr socket) : socket_(socket) {}
+
+    // Network::ListenSocketFactory
+    Network::Address::SocketType socketType() const override { return socket_->socketType(); }
+
+    const Network::Address::InstanceConstSharedPtr& localAddress() const override {
+      return socket_->localAddress();
+    }
+
+    Network::SocketSharedPtr createListenSocket() override { return socket_; }
```
Comment (mattklein123) [inline source/server/http/admin.h]: Given this should only be called once, can you either just create the listen socket inside here and return it without a member variable, or set it to null after this call and then assert it's non-null on function entry?

File: source/server/listener_impl.cc [hunk 9]
```diff
@@ -26,14 +26,83 @@
...
+    // set in the worker after listen()/evconnlistener_new() is called.
+    socket->addOptions(options_);
+  }
+  return socket;
+}
+
+void ListenSocketFactoryImplBase::setLocalAddress(
+    Network::Address::InstanceConstSharedPtr local_address) {
+
+  ENVOY_LOG(info, "Set socket factory local address to {}", local_address->asString());
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: Do we need this debug log? It's unclear what listener this applies to (maybe add name) so I would just delete it, or definitely lower the level to debug or trace.

File: source/server/listener_impl.h [hunk 10]
```diff
@@ -18,6 +19,70 @@ namespace Server {
...
+  // Network::ListenSocketFactory
+  // If |socket_| is nullptr, create a new socket for it. Otherwise, always return |socket_|.
+  Network::SocketSharedPtr createListenSocket() override;
+  absl::optional<std::reference_wrapper<Network::Socket>> sharedSocket() const override {
+    ASSERT(socket_ != nullptr);
+    return *socket_;
+  }
+
+private:
+  // This is shared across all workers if not setting SO_REUSEPORT.
```
Comment (mattklein123) [inline source/server/listener_impl.h]:
```suggestion
  // This is currently always shared across all workers. In the future SO_REUSEPORT support will be added.
```
(Or similar)

File: source/server/listener_impl.h (hunk 4 above)
Comment (l8huang) [reply source/server/listener_impl.h]: I will push a follow up PR for TCP with SO_REUSEPORT.

File: source/server/listener_manager_impl.cc [hunk 11]
```diff
@@ -679,14 +691,21 @@ void ListenerManagerImpl::stopListeners(StopListenersType stop_listeners_type) {
...
+                       // Each listener has its own socket and closes the socket
+                       // on its own.
+                       return;
+                     }
+                     for (auto& listener : active_listeners_) {
+                       if (listener->listenerTag() == listener_tag) {
+                         listener->listenSocketFactory().sharedSocket()->get().close();
+                       }
+                     }
+                     stats_.listener_stopped_.inc();
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: good catch!

File: source/common/event/dispatcher_impl.cc [hunk 12]
```diff
@@ -134,17 +134,17 @@ Filesystem::WatcherPtr DispatcherImpl::createFilesystemWatcher() {
   return Filesystem::WatcherPtr{new Filesystem::WatcherImpl(*this)};
 }
 
-Network::ListenerPtr DispatcherImpl::createListener(Network::Socket& socket,
+Network::ListenerPtr DispatcherImpl::createListener(Network::SocketSharedPtr&& socket,
                                                     Network::ListenerCallbacks& cb,
                                                     bool bind_to_port) {
   ASSERT(isThreadSafe());
-  return std::make_unique<Network::ListenerImpl>(*this, socket, cb, bind_to_port);
+  return std::make_unique<Network::ListenerImpl>(*this, std::move(socket), cb, bind_to_port);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/event/dispatcher_impl.cc]: nope, the passed-in parameter is an rvalue.

File: source/common/network/base_listener_impl.cc [hunk 13]
```diff
@@ -18,14 +18,14 @@ Address::InstanceConstSharedPtr BaseListenerImpl::getLocalAddress(int fd) {
   return Address::addressFromFd(fd);
 }
 
-BaseListenerImpl::BaseListenerImpl(Event::DispatcherImpl& dispatcher, Socket& socket)
-    : local_address_(nullptr), dispatcher_(dispatcher), socket_(socket) {
-  const auto ip = socket.localAddress()->ip();
+BaseListenerImpl::BaseListenerImpl(Event::DispatcherImpl& dispatcher, SocketSharedPtr socket)
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/network/base_listener_impl.cc]: std::shared_ptr has rvalue constructor which avoid increment and decrement of reference counter. Changing to rvalue reference will require many changes in dispatcher interface and tests and doesn't improve performance.

File: source/common/network/listener_impl.cc [hunk 14]
```diff
@@ -64,11 +64,11 @@ void ListenerImpl::setupServerSocket(Event::DispatcherImpl& dispatcher, Socket&
   evconnlistener_set_error_cb(listener_.get(), errorCallback);
 }
 
-ListenerImpl::ListenerImpl(Event::DispatcherImpl& dispatcher, Socket& socket, ListenerCallbacks& cb,
-                           bool bind_to_port)
-    : BaseListenerImpl(dispatcher, socket), cb_(cb), listener_(nullptr) {
+ListenerImpl::ListenerImpl(Event::DispatcherImpl& dispatcher, SocketSharedPtr socket,
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/network/listener_impl.cc]: see my previous comment

File: source/common/network/udp_listener_impl.cc [hunk 15]
```diff
@@ -91,20 +91,20 @@ void UdpListenerImpl::processPacket(Address::InstanceConstSharedPtr local_addres
...
-  return socket_.localAddress();
+  return socket_->localAddress();
 }
 
 Api::IoCallUint64Result UdpListenerImpl::send(const UdpSendData& send_data) {
   ENVOY_UDP_LOG(trace, "send");
   Buffer::Instance& buffer = send_data.buffer_;
   Api::IoCallUint64Result send_result = Utility::writeToSocket(
-      socket_.ioHandle(), buffer, send_data.local_ip_, send_data.peer_address_);
+      socket_->ioHandle(), buffer, send_data.local_ip_, send_data.peer_address_);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/network/udp_listener_impl.cc]: Can you be more specific about the shared_ptr cost here? Most places just de-reference the pointer and member access. Keeping a shared_ptr and a reference for the same instance seems a bit weird to me.

File: source/extensions/quic_listeners/quiche/active_quic_listener.cc [hunk 16]
```diff
@@ -15,27 +15,29 @@ ActiveQuicListener::ActiveQuicListener(Event::Dispatcher& dispatcher,
...
                                        const quic::QuicConfig& quic_config)
     : ActiveQuicListener(dispatcher, parent,
-                         dispatcher.createUdpListener(listener_config.socket(), *this),
+                         listener_config.listenSocketFactory().createListenSocket(),
                          listener_config, quic_config) {}
 
 ActiveQuicListener::ActiveQuicListener(Event::Dispatcher& dispatcher,
                                        Network::ConnectionHandler& parent,
-                                       Network::UdpListenerPtr&& listener,
+                                       Network::SocketSharedPtr listen_socket,
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/active_quic_listener.cc]: see my previous comment

File: source/server/connection_handler_impl.cc [hunk 17]
```diff
@@ -103,7 +101,9 @@ ConnectionHandlerImpl::ActiveListenerImplBase::ActiveListenerImplBase(
 ConnectionHandlerImpl::ActiveTcpListener::ActiveTcpListener(ConnectionHandlerImpl& parent,
                                                             Network::ListenerConfig& config)
     : ActiveTcpListener(
-          parent, parent.dispatcher_.createListener(config.socket(), *this, config.bindToPort()),
+          parent,
+          parent.dispatcher_.createListener(config.listenSocketFactory().createListenSocket(),
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/connection_handler_impl.cc]: My idea is that we always call createListenSocket() here regardless of the socket factory implementation. While in the socket factory, if REUSEPORT is enabled, createListenerSocket() should create a socket each time being called. Or REUSEPORT socket factory can be a stand alone implementation in addition to non-REUSEPORT TCP socket factory.

File: source/server/http/admin.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/http/admin.h]: From class user's perspective, this factory always provides a socket to listen. Whether this socket is newly created or shared should be hidden.

File: source/server/listener_impl.cc [hunk 18]
```diff
@@ -26,14 +26,81 @@
 namespace Envoy {
 namespace Server {
 
+ListenSocketFactoryImplBase::ListenSocketFactoryImplBase(
+    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
+    Network::Address::SocketType socket_type, const Network::Socket::OptionsSharedPtr& options,
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_impl.cc]: This socket factory is a member of ListenerConfig, I would like not to reference to its parent because the time when this socket factory is created, ListenerConfig itself might not be fully populated. And only listener_name  is from ListenerConfig, the rest parameters are temporary variable in the context.

File: source/server/listener_impl.h (hunk 4 above)
Comment (danzh2010) [reply source/server/listener_impl.h]: SG. Might be a shared interface between TCP and UDP socke factory as UDP for sure needs REUSEPORT to scale up.

File: source/server/listener_impl.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_impl.cc]: yes, ListenerConfig doesn't have all params needed right now.

>/**
> * ListenSocketFactory is used to create listening socket for each listener.
> * Listeners created from the same ListenConfig instance share the same
> * ListenSocketFactory instance.
> */

According to `ListenSocketFactory`'s doc string, looks like it makes more sense to use ListenerConfig to construct a ListenSocketFactory. The missed parameters can be provided by ListenerConfig by adding new APIs to it.

File: source/server/http/admin.h (hunk 8 above)
Comment (mattklein123) [reply source/server/http/admin.h]: Sure sounds good.

File: include/envoy/network/listener.h [hunk 19]
```diff
@@ -16,6 +16,43 @@ namespace Network {
...
+
+  /**
+   * @return the type of the socket createListenSocket() returns.
+   */
+  virtual Address::SocketType socketType() const PURE;
+
+  /**
+   * @return the listening address of the socket createListenSocket() returns. Before
+   *         createListenSocket() is called, the return value might has 0 as port number if the
+   * config doesn't specify it.
```
Comment (mattklein123) [inline include/envoy/network/listener.h]: nit: still strange alignment

File: source/common/network/base_listener_impl.h [hunk 20]
```diff
@@ -16,14 +16,18 @@ namespace Network {
...
+   */
+  BaseListenerImpl(Event::DispatcherImpl& dispatcher, SocketSharedPtr socket);
 
 protected:
   virtual Address::InstanceConstSharedPtr getLocalAddress(int fd);
 
   Address::InstanceConstSharedPtr local_address_;
   Event::DispatcherImpl& dispatcher_;
-  Socket& socket_;
+  SocketSharedPtr socket_;
```
Comment (mattklein123) [inline source/common/network/base_listener_impl.h]: nit: const

File: source/common/network/base_listener_impl.h [hunk 21]
```diff
@@ -16,14 +16,18 @@ namespace Network {
  */
 class BaseListenerImpl : public virtual Listener {
 public:
-  BaseListenerImpl(Event::DispatcherImpl& dispatcher, Socket& socket);
+  /**
+   * @param socket the listening socket for this listener. It might be shared
+   * with other listeners if all listeners use single listen socket.
+   */
+  BaseListenerImpl(Event::DispatcherImpl& dispatcher, SocketSharedPtr socket);
```
Comment (mattklein123) [inline source/common/network/base_listener_impl.h]: nit: passing `SocketSharedPtr&&` would be a bit more clear. Same in similar cases. That would likely satisfy @l8huang's readability concern also.

File: source/common/network/udp_listener_impl.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/network/udp_listener_impl.cc]: There shouldn't be any cost to a de-reference as long as there is no inc/dec.

File: source/server/listener_impl.h (hunk 10 above)
Comment (mattklein123) [reply source/server/listener_impl.h]: ping

File: source/server/listener_manager_impl.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: You can just do this at the top of the function and avoid the multiple calls (and yes nice catch).

File: test/server/listener_manager_impl_test.cc [hunk 22]
```diff
@@ -199,6 +199,10 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, DEPRECATED_FEATURE_TEST(TlsContex
 }
 
 TEST_F(ListenerManagerImplWithRealFiltersTest, UdpAddress) {
+
```
Comment (mattklein123) [inline test/server/listener_manager_impl_test.cc]: nit: extra newline

File: source/common/network/udp_listener_impl.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/common/network/udp_listener_impl.cc]: my bad, there are no function call after inline

File: source/common/network/base_listener_impl.h (hunk 21 above)
Comment (mattklein123) [reply source/common/network/base_listener_impl.h]: OK that's fine.
